include/sunlinsol/sunlinsol_ginkgobatch.hpp

```diff
@@ -0,0 +1,505 @@
+  BatchLinearSolver(std::shared_ptr<const gko::Executor> gko_exec,
+                    gko::batch::stop::tolerance_type tolerance_type,
+                    int max_iters, gko::size_type num_batches, SUNContext sunctx)
+    : BatchLinearSolver(std::move(gko_exec), tolerance_type, nullptr, max_iters,
+                        num_batches, sunctx)
+  {}
+
+  // NOLINTNEXTLINE(cppcoreguidelines-pro-type-member-init)
+  BatchLinearSolver(std::shared_ptr<const gko::Executor> gko_exec,
+                    std::shared_ptr<gko::batch::BatchLinOpFactory> precon_factory,
+                    int max_iters, gko::size_type num_batches, SUNContext sunctx)
+    : BatchLinearSolver(std::move(gko_exec),
+                        gko::batch::stop::tolerance_type::absolute,
+                        precon_factory, max_iters, num_batches, sunctx)
+  {}
+
+  // NOLINTNEXTLINE(cppcoreguidelines-pro-type-member-init)
+  BatchLinearSolver(std::shared_ptr<const gko::Executor> gko_exec,
+                    gko::batch::stop::tolerance_type tolerance_type,
+                    std::shared_ptr<gko::batch::BatchLinOpFactory> precon_factory,
+                    gko::size_type num_batches, SUNContext sunctx)
+    : BatchLinearSolver(std::move(gko_exec), tolerance_type, precon_factory,
+                        default_max_iters_, num_batches, sunctx)
+  {}
+
+  /// Get the ``gko::Executor`` associated with the Ginkgo matrix
+  std::shared_ptr<const gko::Executor> GkoExec() const { return gko_exec_; }
+
+  // Override the ConvertibleTo methods
+
+  /// Implicit conversion to a :c:type:`SUNLinearSolver`
+  operator SUNLinearSolver() override { return object_.get(); }
+
+  /// Implicit conversion to a :c:type:`SUNLinearSolver`
+  operator SUNLinearSolver() const override { return object_.get(); }
+
+  /// Explicit conversion to a :c:type:`SUNLinearSolver`
+  SUNLinearSolver Convert() override { return object_.get(); }
+
+  /// Explicit conversion to a :c:type:`SUNLinearSolver`
+  SUNLinearSolver Convert() const override { return object_.get(); }
+
+  /// Get the underlying Ginkgo solver
+  /// \note This will be `nullptr` until the linear solver setup phase.
+  GkoBatchSolverType* GkoSolver() { return solver_.get(); }
+
+  /// Average number of iterations across the batches during the last solve.
+  sunrealtype AvgNumIters() const { return avg_iter_count_; }
+
+  /// Standard deviation of the number of iterations across the batches during the last solve.
+  sunrealtype StddevNumIters() const { return stddev_iter_count_; }
+
+  /// Running sum of the average number of iterations in this solvers lifetime.
+  sunrealtype SumAvgNumIters() const { return sum_of_avg_iters_; }
+
+  void SetScalingMode(int scaling_mode = LAGGED_SCALING)
+  {
+    scaling_mode_ = scaling_mode;
+  }
+
+  /// Sets the left and right scaling vectors to be used.
+  void SetScalingVectors(N_Vector s1, N_Vector s2)
+  {
+    SUNFunctionBegin(sunCtx());
+
+    if (s1 && s2)
+    {
+      s1_ = s1;
```

Comment:
I don't think we would change `s1` between the setup and solve calls but, if for some reason we did, this should be a deep copy (and we'd need to allocate a vector for `s1_` which could be conditional on the scaling type used).